List assignment parsing error

Code below doesn't works:

```sh
a = "a"
b = ($a + "b")
ERROR: List assignment not finished
```

```sh
	reqdirs = (
		$mnt+"/proc"
		$mnt+"/sys"
		$mnt+"/dev"
		$mnt+"/tmp"
		$mnt+"/run"
		$mnt+"/var/cache/pacman/pkg"
		$mnt+"/var/lib/pacman"
		$mnt+"/var/log"
	)
[ERROR] nashfmt:42:7: List assignment not finished
```

--- a/ast/tree.go
+++ b/ast/tree.go
@@ -38,9 +38,31 @@ func (tree *Tree) String() string {
 
 	content := make([]string, 0, 8192)
 
+	isAssigns := false
+
 	for i := 0; i < len(nodes); i++ {
 		node := nodes[i]
+
 		nodebytes := node.String()
+
+		if i == 0 && node.Type() == NodeComment {
+			nodebytes += "\n"
+		}
+
+		if i < (len(nodes) - 1) {
+			nextNode := nodes[i+1]
+
+			switch nextNode.Type() {
+			case NodeAssignment:
+				isAssigns = true
+			default:
+				if isAssigns {
+					nodebytes += "\n"
+				}
+
+				isAssigns = false
+			}
+		}
 
 		content = append(content, nodebytes)
 	}

--- a/errors/error.go
+++ b/errors/error.go
@@ -1,6 +1,10 @@
 package errors
 
-import "fmt"
+import (
+	"fmt"
+
+	"github.com/NeowayLabs/nash/scanner"
+)
 
 type (
 	NashError struct {
@@ -29,17 +33,17 @@ func (e *NashError) SetReason(format string, arg ...interface{}) {
 
 func (e *NashError) Error() string { return e.reason }
 
-func NewUnfinishedBlockError() error {
+func NewUnfinishedBlockError(name string, it scanner.Token) error {
 	return &unfinishedBlockError{
-		NashError: NewError("Statement's block '{' not finished"),
+		NashError: NewError("%s:%d:%d: Statement's block '{' not finished", name, it.Line(), it.Column()),
 	}
 }
 
 func (e *unfinishedBlockError) Unfinished() bool { return true }
 
-func NewUnfinishedListError() error {
+func NewUnfinishedListError(name string, it scanner.Token) error {
 	return &unfinishedListError{
-		NashError: NewError("List assignment not finished"),
+		NashError: NewError("%s:%d:%d: List assignment not finished", name, it.Line(), it.Column()),
 	}
 }
 

--- a/parser/parse.go
+++ b/parser/parse.go
@@ -545,7 +545,7 @@ func (p *Parser) parseAssignValue(name scanner.Token) (ast.Node, error) {
 		it = p.peek()
 
 		for it.Type() == token.Arg || it.Type() == token.String || it.Type() == token.Variable {
-			arg, err := p.getArgument(true, false)
+			arg, err := p.getArgument(true, true)
 
 			if err != nil {
 				return nil, err
@@ -557,7 +557,7 @@ func (p *Parser) parseAssignValue(name scanner.Token) (ast.Node, error) {
 		}
 
 		if it.Type() != token.RParen {
-			return nil, errors.NewUnfinishedListError()
+			return nil, errors.NewUnfinishedListError(p.name, it)
 		}
 
 		p.next()
@@ -935,7 +935,7 @@ func (p *Parser) parseReturn() (ast.Node, error) {
 		}
 
 		if valueIt.Type() != token.RParen {
-			return nil, errors.NewUnfinishedListError()
+			return nil, errors.NewUnfinishedListError(p.name, valueIt)
 		}
 
 		p.next()
@@ -1065,7 +1065,7 @@ func (p *Parser) parseBlock() (*ast.ListNode, error) {
 
 finish:
 	if p.openblocks != 0 {
-		return nil, errors.NewUnfinishedBlockError()
+		return nil, errors.NewUnfinishedBlockError(p.name, p.peek())
 	}
 
 	return ln, nil

--- a/parser/parse_fmt_test.go
+++ b/parser/parse_fmt_test.go
@@ -88,3 +88,31 @@ testb = "b"`},
 
 	testFmtTable(testTable, t)
 }
+
+func TestFmtGroupVariables(t *testing.T) {
+	testTable := []fmtTestTable{
+		{
+			`test = "a"
+
+test2 = "b"
+
+fn cd() { echo "hello" }`,
+			`test = "a"
+test2 = "b"
+
+fn cd() {
+	echo "hello"
+}
+`,
+		},
+		{
+			`#!/usr/bin/env nash
+echo "hello"`,
+			`#!/usr/bin/env nash
+
+echo "hello"`,
+		},
+	}
+
+	testFmtTable(testTable, t)
+}

--- a/parser/parse_regression_test.go
+++ b/parser/parse_regression_test.go
@@ -119,3 +119,26 @@ func TestParseIssue68(t *testing.T) {
 
 	parserTestTable("parse issue #68", `cat PKGBUILD | sed "s#\\$pkgdir#/home/i4k/alt#g" > PKGBUILD2`, expected, t, false)
 }
+
+func TestParseIssue69(t *testing.T) {
+	expected := ast.NewTree("parse-issue-69")
+	ln := ast.NewListNode()
+
+	parts := make([]ast.Expr, 2)
+
+	parts[0] = ast.NewVarExpr(0, "$a")
+	parts[1] = ast.NewStringExpr(0, "b", true)
+
+	concat := ast.NewConcatExpr(0, parts)
+
+	listValues := make([]ast.Expr, 1)
+	listValues[0] = concat
+
+	list := ast.NewListExpr(0, listValues)
+
+	assign := ast.NewAssignmentNode(0, "a", list)
+	ln.Push(assign)
+	expected.Root = ln
+
+	parserTestTable("parse-issue-69", `a = ($a+"b")`, expected, t, true)
+}

--- a/parser/parse_test.go
+++ b/parser/parse_test.go
@@ -232,6 +232,7 @@ func TestParseWithShebang(t *testing.T) {
 	expected.Root = ln
 
 	parserTestTable("parser shebang", `#!/bin/nash
+
 echo bleh
 `, expected, t, true)
 }

--- a/scanner/lex.go
+++ b/scanner/lex.go
@@ -555,7 +555,7 @@ func lexInsideListVariable(l *Lexer) stateFn {
 		return lexStart
 	}
 
-	return l.errorf("Unexpected '%q'. Expected elements or ')'", r)
+	return lexStart
 }
 
 func lexInsideCommonVariable(l *Lexer, nextConcatFn stateFn, nextFn stateFn) stateFn {

--- a/scanner/lex_regression_test.go
+++ b/scanner/lex_regression_test.go
@@ -402,3 +402,19 @@ func TestLexerIssue85(t *testing.T) {
 
 	testTable("test issue 85", `a <= -echo hello`, expected, t)
 }
+
+func TestLexerIssue69(t *testing.T) {
+	expected := []Token{
+		{typ: token.Ident, val: "a"},
+		{typ: token.Assign, val: "="},
+		{typ: token.LParen, val: "("},
+		{typ: token.Variable, val: "$a"},
+		{typ: token.Concat, val: "+"},
+		{typ: token.String, val: "b"},
+		{typ: token.RParen, val: ")"},
+		{typ: token.EOF},
+	}
+
+	testTable("test69", `a = ($a + "b")`, expected, t)
+
+}

--- a/scanner/lex_test.go
+++ b/scanner/lex_test.go
@@ -37,6 +37,12 @@ func testTable(name, content string, expected []Token, t *testing.T) {
 			fmt.Printf("parsed: %+v\n", res)
 		}
 
+		fmt.Printf("\n")
+
+		for _, exp := range expected {
+			fmt.Printf("expect: %+v\n", exp)
+		}
+
 		return
 	}
 

